### PR TITLE
Player 1340

### PR DESCRIPTION
**Pull request by pilievOoyala**

### CODE REVIEW

**Comment by pilievOoyala** [comment]
I figured out another unit test I should add. Let me do that before we merge.

File: js/google_ima.js [hunk 1]
```diff
@@ -330,6 +336,22 @@ require("../html5-common/js/utils/utils.js");
         }
       };
 
+      var _setNonAdRulesPreroll = privateMember(function(validAdTags)
```
Comment (aeng7) [inline js/google_ima.js]: Can we include a jsdoc for this?

File: js/google_ima.js [hunk 2]
```diff
@@ -330,6 +336,22 @@ require("../html5-common/js/utils/utils.js");
...
+      {
+        if (!_usingAdRules)
+        {
+          for (var index = 0; index < validAdTags.length; index++)
+          {
+            if (validAdTags[index].position == 0)
```
Comment (aeng7) [inline js/google_ima.js]: Can we use === here instead of ==?

File: js/google_ima.js [hunk 3]
```diff
@@ -330,6 +336,22 @@ require("../html5-common/js/utils/utils.js");
         }
       };
 
+      var _setNonAdRulesPreroll = privateMember(function(validAdTags)
+      {
+        if (!_usingAdRules)
```
Comment (aeng7) [inline js/google_ima.js]: Can we add a null check for validAdTags?

File: test/unit-tests/ima_test.js [hunk 4]
```diff
@@ -1839,4 +1839,20 @@ describe('ad_manager_ima', function()
       expect(_.isEmpty(sdkAdEventParams)).to.be(false);
     }
   });
+
+  it('IMA ad preloading: testing page level param', function()
```
Comment (aeng7) [inline test/unit-tests/ima_test.js]: Do we have unit tests for the preloading (early ad request) functionality already? If not, can we add some?

File: js/google_ima.js [hunk 5]
```diff
@@ -536,10 +559,14 @@ require("../html5-common/js/utils/utils.js");
           //if we are trying to play an linear ad then we need to request the ad now.
           if (this.currentAMCAdPod.ad.forced_ad_type != _amc.ADTYPE.NONLINEAR_OVERLAY)
           {
-            //reset adRequested and adTagUrl so we can request another ad
-            _resetAdsState();
-            this.adTagUrl = this.currentAMCAdPod.ad.tag_url;
-            _trySetupAdsRequest();
+            //if this is the preroll we preloaded then don't try to request it again.
+            if (!(this.preloadAds && this.numAdsPlayed === 1 && this.currentAMCAdPod.position === 0))
```
Comment (aeng7) [inline js/google_ima.js]: Can you explain this condition to me?

Comment (pilievOoyala) [inline js/google_ima.js]: I can't sneak anything past you :).  With ad preloading, we only preloads the first preroll if it exists.  So this is checking if the first ad to play is a preroll and we are preloading, then we have already requested the ad, so skip requesting it again.

Now that I've explained it, I probably need to go back and figure out a more generic way to mark ads in the AMC as being preloaded and then we just check a boolean.  Then we can reuse this for anything.

**Pull request closed without merging**